karabo_bridge/simulation.py

```diff
@@ -29,125 +29,183 @@
-    }
-}
-
-
-def gen_combined_detector_data(detector_info, tid_counter, corrected=False,
-                               nsources=1):
-    source = detector_info['source']
-    gen = {source: {}}
-    meta = {}
-
-    # metadata
-    ts = time()
-    sec, frac = str(ts).split('.')
-    meta[source] = {
-        'source': source,
-        'timestamp': ts,
-        'timestamp.tid': tid_counter,
-        'timestamp.sec': sec,
-        'timestamp.frac': frac.ljust(18, '0')  # attosecond resolution
-    }
+class Detector:
+    # Overridden in subclasses
+    pulses = 0  # nimages per XRAY train
+    modules = 0  # nb of super modules composing the detector
+    mod_x = 0  # pixel count (y axis) of a single super module
+    mod_y = 0  # pixel count (x axis) of a single super module
+    pixel_size = 0  # [mm]
+    distance = 0  # Sample to detector distance [mm]
+    layout = np.array([[]])  # super module layout of the detector
+    const_fields = {}
+
+    @staticmethod
+    def getDetector(detector, source='', corr=True, gen='random'):
+        if detector == 'AGIPD':
+            source = source or 'SPB_DET_AGIPD1M-1/DET/detector'
+            return AGIPD(source, corr=corr, gen=gen)
+        elif detector == 'LPD':
+            source = source or 'FXE_DET_LPD1M-1/DET/detector'
+            return LPD(source, corr=corr, gen=gen)
+        else:
+            raise NotImplementedError('detector %r not available' % detector)
+
+    def __init__(self, source='', corr=False, gen='random'):
+        self.source = source or 'INST_DET_GENERIC/DET/detector'
+        self.corrected = corr
+        if gen == 'random':
+            self.genfunc = self.random
+        elif gen == 'zeros':
+            self.genfunc = self.zeros
+        else:
+            raise NotImplementedError('gen func %r not implemented' % gen)
+
+    @property
+    def data_shape(self):
+        return (self.modules, self.mod_y, self.mod_x, self.pulses)
 
-    pulse_count = detector_info['pulses']
-    array_shape = tuple((detector_info['modules'], ) +
-                        detector_info['module_shape'] + (pulse_count, ))
+    @property
+    def data_type(self):
+        return np.float32 if self.corrected else np.uint16
 
-    # detector random data
-    if corrected:
-        gain_data = np.zeros(array_shape, dtype=np.uint16)
-        domain = detector_info['source'].partition('/')[0]
+    def corr_passport(self):
+        domain = self.source.partition('/')[0]
         passport = [
             '%s/CAL/THRESHOLDING_Q1M1' % domain,
             '%s/CAL/OFFSET_CORR_Q1M1' % domain,
             '%s/CAL/RELGAIN_CORR_Q1M1' % domain
         ]
+        return passport
+
+    def random(self):
+        return np.random.uniform(low=1500, high=1600,
+                                 size=self.data_shape).astype(self.data_type)
+
+    def zeros(self):
+        return np.zeros(self.data_shape, dtype=self.data_type)
+
+    def module_position(self, ix):
```

Comment:
What's this for? It doesn't look like it's used anywhere. Is it there for adding more complex simulated patterns in the future? Will this simple grid layout be enough, or should we look at positioning the modules based on some kind of (possibly generated) geometry?

Reply:
it is not use yet but you got the idea. I think we would need a geometry / module layout + x,y direction for each modules to be able to use this.